### PR TITLE
Add Magento authentication script.

### PR DESCRIPTION
**Pull request by Radio**

### CODE REVIEW

File: authentication/MagentoAuthentication.js [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+    
+    var loginUri = new URI(paramsValues.get("loginUrl"), false);
+    
+    // Perform a GET request to the login page to get the form_key
+    var get = helper.prepareMessage();
+    get.setRequestHeader(new HttpRequestHeader(HttpRequestHeader.GET, loginUri, HttpHeader.HTTP10));
+    helper.sendAndReceive(get);
+    var formParameters = parseFormParameters(get.getResponseBody().toString());
+    
+    // Build the request body using the credentials values and the CAS values obtained from the first request
```
Comment (kingthorin) [inline authentication/MagentoAuthentication.js]: Looks like this comment needs updating, guess it's a hold over from the source/template.
> CAS values....

File: authentication/MagentoAuthentication.js [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+    // Enable Rhino behavior, in case ZAP is running on Java 8 (which uses Nashorn)
+    if (java.lang.System.getProperty("java.version").startsWith("1.8")) {
+        load("nashorn:mozilla_compat.js");
+    }
+    
+    // Imports
+    importClass(org.parosproxy.paros.network.HttpRequestHeader);
+    importClass(org.parosproxy.paros.network.HttpHeader);
+    importClass(org.apache.commons.httpclient.URI);
+    importClass(org.apache.commons.httpclient.params.HttpClientParams);
```
Comment (thc202) [inline authentication/MagentoAuthentication.js]: Could be removed, not used.

File: authentication/MagentoAuthentication.js [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+    var regex = "<input.*name=\"(form_key)\".*value=\"([^\"]*)\"";
+    var matcher = Pattern.compile(regex).matcher(response);
+    while (matcher.find()) {
+        result[matcher.group(1)] = matcher.group(2);
+    }
+    
+    return result;
+}
+
+function getHttpClientFromHelper(helper) {
```
Comment (thc202) [inline authentication/MagentoAuthentication.js]: Could be removed, not used.

File: authentication/MagentoAuthentication.js [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+function getOptionalParamsNames(){
+    return ["extraPostData"];
+}
+
+function getCredentialsParamsNames(){
+    return ["username", "password"];
+}
+
+function authenticate(helper, paramsValues, credentials) {
+    print("---- Magento authentication script has started ----");
```
Comment (thc202) [inline authentication/MagentoAuthentication.js]: These helper messages could be wrapped in a debug flag (default to false?).

Comment (Radio) [inline authentication/MagentoAuthentication.js]: Like this?
```
 debugMode && print("...");
```
Do you mean just a local variable or a parameter from UI? Or is there already a built-in debug mode parameter?

Comment (thc202) [inline authentication/MagentoAuthentication.js]: Yeah, that's fine. A local variable (ZAP does not provide anything to control that).

Comment (kingthorin) [inline authentication/MagentoAuthentication.js]: Just create one ... maybe after the imports.
``` Javascript 
var debug = false;
```

Then something like:
```Javascript
if (debug == true) {
    print("---- Magento authentication script has finished ----\n");
}
```

File: authentication/MagentoAuthentication.js (hunk 1 above)
Comment (Radio) [inline authentication/MagentoAuthentication.js]: Thanks, fixed.

File: authentication/MagentoAuthentication.js (hunk 2 above)
Comment (Radio) [inline authentication/MagentoAuthentication.js]: Thanks, removed.

File: authentication/MagentoAuthentication.js (hunk 3 above)
Comment (Radio) [inline authentication/MagentoAuthentication.js]: Thanks, removed.

File: authentication/MagentoAuthentication.js (hunk 4 above)
Comment (Radio) [inline authentication/MagentoAuthentication.js]: Thx, done.

**Comment by Radio** [comment]
Rebased into one, please review.

**Comment by thc202** [comment]
Thanks!

**Comment by kingthorin** [comment]
Thanks @Radio

**Pull request merged**